server/src/main/java/org/elasticsearch/cluster/coordination/StableMasterHealthIndicatorService.java

```diff
@@ -0,0 +1,202 @@
+                DiscoveryNode master = localMasterHistory.getMostRecentNonNullMaster();
+                logger.trace("One master has gone null 4 or more times recently: " + master);
+                boolean localNodeIsMaster = clusterService.localNode().equals(master);
+                final List<DiscoveryNode> remoteHistory;
+                if (localNodeIsMaster) {
+                    remoteHistory = null; // We don't need to fetch the remote master's history if we are that remote master
+                } else {
+                    remoteHistory = masterHistoryService.getRemoteMasterHistory(master);
+                }
+                if (localNodeIsMaster || remoteHistory == null || getNumberOfMasterChanges(remoteHistory) > 3) {
+                    if (localNodeIsMaster == false && remoteHistory == null) {
+                        logger.trace("Unable to get master history from {}}", master);
+                    } else {
+                        logger.trace("The master node {} thinks it is unstable", master);
+                    }
+                    stableMasterStatus = HealthStatus.YELLOW;
+                    summary = String.format(
+                        Locale.ROOT,
+                        "The cluster's master has alternated between %s and no master multiple times in the last 30 minutes",
+                        master
+                    );
+                    impacts.add(
+                        new HealthIndicatorImpact(
+                            3,
+                            "The cluster is at risk of not being able to create, delete, or rebalance indices",
+                            List.of(ImpactArea.INGEST)
+                        )
+                    );
+                    if (includeDetails) {
+                        details.put("current_master", new DiscoveryNodeXContentObject(localMasterHistory.getCurrentMaster()));
+                    }
+                } else {
+                    logger.trace("This node thinks the master is unstable, but the master node {} thinks it is stable", master);
+                    stableMasterStatus = HealthStatus.GREEN;
+                    summary = "The cluster has a stable master node";
+                }
+            } else {
+                logger.trace("The cluster has a stable master node");
+                stableMasterStatus = HealthStatus.GREEN;
+                summary = "The cluster has a stable master node";
+            }
+        } else {
+            stableMasterStatus = HealthStatus.RED;
+            summary = "Placeholder summary";
+        }
+
+        return createIndicator(
+            stableMasterStatus,
+            summary,
+            includeDetails ? new SimpleHealthIndicatorDetails(details) : HealthIndicatorDetails.EMPTY,
+            impacts
+        );
+    }
+
+    long getNumberOfMasterChanges(MasterHistory localMasterHistory) {
+        return getNumberOfMasterChanges(localMasterHistory.getImmutableView());
+    }
```

Comment:
This is only called once outside of tests, we could probably make `getNumberOfMasterChanges(List<DiscoveryNode> masterHistory)` more visible (and `static`) and then remove this method.

Reply:
Done.